### PR TITLE
[build-script] Add support for using isolated -impl actions.

### PR DESCRIPTION
**Pull request by ddunbar**
#### What's in this pull request?

This adopts the feature in #2880 in `build-script`, and is part of SR-237.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

- This change moves the top-level invocation driver loop into `build-script`
  and uses the `-impl` script to perform each individual action. Once landed
  and enabled, this will enable us to migrate the individual pieces of the
  `-impl` script into Python code in an incremental fashion.
- This also introduces stub product definitions for each of the different projects
  we manage.
- This works, but is disabled by default (`--no-legacy-impl`) because it
  severely impacts the performance of null builds (4x slower, currently) due to
  the `build-script-impl` parsing overhead. If only we had a JITing bash
  implementation...

### CODE REVIEW

**Comment by ddunbar** [comment]
@swift-ci please smoke test

**Comment by ddunbar** [comment]
@swift-ci please python lint

**Comment by ddunbar** [comment]
@swift-ci please python lint

**Comment by ddunbar** [comment]
@swift-ci please smoke test

File: utils/build-script [hunk 1]
```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
...
+                assert name == "package", "invalid action"
+                action_name = "{}-{}".format(host.name, name)
+            else:
+                assert name is not None, "invalid action"
+                action_name = "{}-{}-{}".format(
+                    host.name, product_class.product_name(), name)
+            call_without_sleeping(
+                [build_script_impl] + impl_args + [
+                    "--only-execute", action_name],
+                env=impl_env, echo=self.args.verbose_build)
```
Comment (ddunbar) [inline utils/build-script]: Note that this mode drops echoing subcommands by default, it is too verbose until we finish killing the `-impl` script.

File: utils/swift_build_support/swift_build_support/products/product.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+
+class Product(object):
+    @classmethod
+    def product_name(klass):
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/products/product.py]: This should be cls according to pep8.

https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments

File: utils/swift_build_support/swift_build_support/products/product.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        """product_name() -> str
+
+        The identifier-style name to use for this product.
+        """
+        return klass.__name__.lower()
+
+    @classmethod
+    def get_build_directory_name(klass, host_target):
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/products/product.py]: Same here.

File: utils/build-script [hunk 4]
```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
 
         return options
 
+    def compute_product_classes(self):
```
Comment (gottesmm) [inline utils/build-script]: Do we really need to introduce a new concept called product classes. It seems like what you are really talking about are projects, no? Why not just make this a property called: "projects".

Comment (gottesmm) [inline utils/build-script]: After reading the rest of the commit, I see what you are doing in terms of product_classes. But still, why not just make it a property called product_classes and then cache the result?

Comment (karwa) [inline utils/build-script]: If I'm not mistaken, "product_classes" refers to the fact that these are classes (rather than instances) of the products.

File: utils/swift_build_support/swift_build_support/products/product.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        return "{}-{}".format(klass.product_name(),
+                              host_target.name)
+
+    def __init__(self, args, toolchain, source_dir, build_dir):
+        self.args = args
```
Comment (karwa) [inline utils/swift_build_support/swift_build_support/products/product.py]: Can we remove the `args` part of this initialiser back to Ninja? I don't like the idea of each product parsing its own options from the build-script; it makes it easy to lose track of which options are required and for what. For and example of where this is already a problem, see `CMake` (cmake.py); it also parses its own build-script arguments, and for reasons I can't explain is setting sanitizers and compiler version strings (both of which only make sense for LLVM/Swift products). I feel that something like that would be easier to spot if `CMake()` took a `clang_compiler_version` argument. Somebody working on the general build-script would realise that it doesn't make sense for the general CMake object to require that information.

Basically, I think we should avoid splitting the argument parsing logic up from one file (`build-script`), to a collection of many files.

File: utils/build-script [hunk 6]
```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
...
+        all_host_names = [
+            self.args.host_target] + self.args.cross_compile_hosts
+        all_hosts = [StdlibDeploymentTarget.get_target_for_name(name)
+                     for name in all_host_names]
+
+        # Compute the list of product classes to operate on.
+        #
+        # FIXME: This should really be per-host, but the current structure
+        # matches that of `build-script-impl`.
+        product_classes = self.compute_product_classes()
```
Comment (karwa) [inline utils/build-script]: The reason it's not per-host is (as usual) because we don't have the build-script argument structure necessary to define a list of products per host; they are defined with flags such as "--foundation".

File: utils/build-script [hunk 7]
```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
...
+        is constructed in dependency order.
+        """
+
+        # FIXME: This is a weird division (returning a list of class objects),
+        # but it matches the existing structure of the `build-script-impl`.
+
+        product_classes = []
+        product_classes.append(products.CMark)
+        product_classes.append(products.LLVM)
+        product_classes.append(products.Clang)
```
Comment (gribozavr) [inline utils/build-script]: Did you mean to use 'Swift' instead of 'Clang'?  LLVM and Clang build in the same invocation.

File: utils/build-script (hunk 1 above)
Comment (karwa) [inline utils/build-script]: Looking at this function, I don't like us storing all build-script arguments (`self.args`) and basically doing our own parsing of them here. I think the instance should request and store the specific properties it requires, and let the build-script worry about fulfilling them from the arguments it received.

Additionally, to keep the options manageable and easy to reason about, I would have the initializers only require non-reasonably-defaultable arguments (e.g. `verbose_build` can reasonably be defaulted to `False`, but we can't just invent `toolchain` or whether or not to use `dist_cc` while preserving the essence of what the user requested). Optional flags like `verbose_build` would be set in a following function.

File: utils/swift_build_support/swift_build_support/products/clang.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+from . import product
+
+
+class Clang(product.Product):
```
Comment (gribozavr) [inline utils/swift_build_support/swift_build_support/products/clang.py]: Do we need separate products for LLVM and Clang?  They are built within the same CMake invocation.

File: utils/build-script [hunk 9]
```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
...
+        # FIXME: This should really be per-host, but the current structure
+        # matches that of `build-script-impl`.
+        product_classes = self.compute_product_classes()
+
+        # Execute each "pass".
+
+        # Build...
+        for host_target in all_hosts:
+            # FIXME: We should only compute these once.
+            config = HostSpecificConfiguration(host_target.name, self)
```
Comment (karwa) [inline utils/build-script]: Yes, BuildScriptInvocation is not very well structured here. It should explicitly calculate the configurations in its initialiser, instead of in `convert_to_impl_args`, and store them as an instance variable. Then iterating over `all_hosts` here just comes an iteration over the `HostSpecificConfiguration`s (Host available via `config.host_target`, Targets available via `config.stdlib_targets_to_configure/build`).

File: utils/build-script (hunk 7 above)
Comment (ddunbar) [inline utils/build-script]: Yup 😊

File: utils/build-script (hunk 4 above)
Comment (ddunbar) [inline utils/build-script]: We will eventually need this to be host-specific, so it doesn't make sense to be a property.

File: utils/swift_build_support/swift_build_support/products/product.py (hunk 5 above)
Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/product.py]: I agree, I don't like this either parsing of arguments outside of the main `build-script`.

I think it would be better to keep the common initializer, though, and instead refactor it to take a more explicit value type instead of the full argument set.

File: utils/swift_build_support/swift_build_support/products/clang.py (hunk 8 above)
Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/clang.py]: No, probably not, I'll pull it out.

File: utils/swift_build_support/swift_build_support/products/product.py (hunk 5 above)
Comment (karwa) [inline utils/swift_build_support/swift_build_support/products/product.py]: I would prefer if the root initialiser didn't take args. For now I reckon the subclass initialisers can still take it, and we can refactor it out later.

Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/product.py]: Yeah, I am agreeing with that, and proposing that instead of taking args, it take a more structured object. I'll move args back to Ninja for now though...

File: utils/build-script (hunk 4 above)
Comment (karwa) [inline utils/build-script]: As I mention in a longer comment below, I think it should be a property on HostSpecificConfiguration. For now we'd set every instance to the same parsed set of products.

File: utils/build-script (hunk 9 above)
Comment (ddunbar) [inline utils/build-script]: I basically agree with all your points, but the problem is that it is very hard to enact cleanups when the implementation is split between the Python code and the bash code. I really think we are better off getting the migration done and then continuing to clean up the code.

More specific comments:

1. The reason I didn't put HostSpecificConfiguration into its own file was because its logic was so intimately tied to the current `args` set (which we agree is insufficient). I didn't want to have a constructor in `swift_build_support` that just used a grab bag of `args` items, and I also didn't want to try and explode all of the things it currently uses into separate arguments (or its own object). I'm not sure how to resolve this difficulty currently, but agree it would be good to have something better and have tests. Do you have ideas?

2. I agree about making everything store more structured objects instead of names, and will get to that (eventually). I'm happy to have the #2835 version of this.

3. I agree we should have an object that represents the parsed arguments better than the existing invocation, and that object should live inside `swift_build_support`. Having `Invocation` be the sole object which understand the "command line arguments" and turns that into structured information makes sense to me though, and matches a convention we use inside the Clang and Swift compilers.

File: utils/swift_build_support/swift_build_support/products/product.py (hunk 2 above)
Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/product.py]: Fixed

File: utils/swift_build_support/swift_build_support/products/product.py (hunk 5 above)
Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/product.py]: I investigated this, and I remember why I didn't do it: I want all the products to have the same initializer signature so that the `build-script` can work with the classes generically. I think that is more important than changing the location of the `args` initializer (given that it already leaks out). Instead I think we should just change the Ninja product to stop requiring `args`.

File: utils/swift_build_support/swift_build_support/products/clang.py (hunk 8 above)
Comment (ddunbar) [inline utils/swift_build_support/swift_build_support/products/clang.py]: Fixed

File: utils/build-script (hunk 7 above)
Comment (ddunbar) [inline utils/build-script]: Fixed

**Comment by ddunbar** [comment]
@swift-ci please smoke test

**Comment by rintaro** [comment]
Continued from: https://github.com/apple/swift/pull/2988#discussion_r66891436

@ddunbar 

> I want all the products to have the same initializer signature so that the build-script can work with the classes generically.

According to my own experience, difficult part for that is "dependency": `Swift` depends on `CMark` and `LLVM`, `Foundation` depends on `Swift`,  and so on.

FYI, Here's my *outdated* attempt from #2121:
https://github.com/rintaro/swift/blob/SR-237-build-script-rewrite/utils-experimental/build-script-impl/build_script/main_driver.py#L321-L483

From an object-oriented perspective, I think, only `LLVM` object [should knows the path of its deliverables](https://github.com/rintaro/swift/blob/SR-237-build-script-rewrite/utils-experimental/build-script-impl/build_script/products/llvm.py#L88-L129).
That means, `Swift` object must somehow reference `LLVM` object for the host, or receive deliverable paths from the main control flow.
Maybe, we can pass structured object that holds all `product` object, but that leads the same problem of `args`: it easy to lose track of which products depends on which.
Or should we invent some dependency manager?

@karwa As for `args` problem

> it makes it easy to lose track of which options are required and for what.

I agree with your concern. But, for example, `Swift` requires over 30 properties of `args`.
Nobody wants to do like this:
```python
swift_obj = products.Swift(
    target=target,

    # args
    native_llvm_tools_path=args.native_llvm_tools_path,
    native_clang_tools_path=args.native_clang_tools_path,
    native_swift_tools_path=args.native_tools_swift_path,
    build_type=args.swift_build_type,
    enable_assertions=args.swift_enable_assertions,
    ...
    ... )
```

Do you have ideas?

**Comment by ddunbar** [comment]
One possible way to manage that, and I'm not familiar enough with argparse to know how easy this is, but if we could get the highly product dependent arguments to simply be parsed into a more structured product-specific object (SwiftProductArgs, e.g.) that would at least be an improvement.

**Comment by rintaro** [comment]
> Is it possible to get an argparse argument dest to map to a subobject?

Like this?
```python
parser = argparse.ArgumentParser()
parser.add_argument('--swift-opt', dest="swift.opt")

args = parser.parse_args(['--swift-opt', 'foobar'])
assert args.swift.opt == 'foobar'
```
It's possible if we use custom `Namespace` object.
Roughly implemented demo:
https://gist.github.com/rintaro/9f99f4b1627aa9daf91e643f7e5e6a7c

**Comment by karwa** [comment]
I don't think it's such a big problem. It's only the Swift project that really takes a lot of parameters. Many of them can be given reasonable defaults for the purpose of keeping the initialiser manageable. The build script would then parse the args and set the appropriate properties.

Many of those arguments could also be folded - e.g. the native_{project}_tools_path arguments could be derived from a native HostSpecificConfiguration argument. Actually, in this case the native toolchain is only required for cross-compile builds, so we can leave it out of the initialiser and default the properties to None. In general, any build settings Swift gets from, e.g, LLVM, should come from an already-parsed LLVM object. Thats also easier to reason about and makes semantic sense.

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by ddunbar** [comment]
@swift-ci please smoke test

**Comment by ddunbar** [comment]
I am going to go ahead and land this change, in the interest of continuing the SR-237 conversion. I'll continue to iterate on the feedback to simplify and cleanup the build-script, but I want to split that into separate PRs.

**Comment by ddunbar** [comment]
@rintaro 

> From an object-oriented perspective, I think, only LLVM object should knows the path of its deliverables.
> That means, Swift object must somehow reference LLVM object for the host, or receive deliverable paths from the main control flow.
> Maybe, we can pass structured object that holds all product object, but that leads the same problem of args: it easy to lose track of which products depends on which

I think the relationships between the products are simple enough that this is a good pragmatic approach, we have one structured object to hold all the product objects, and pass that to each product so that it can get connection information from that.

**Comment by karwa** [comment]
I would say the opposite: our dependencies are so simple they can be passed manually.

Especially if we split the build step in to tools/target libraries steps as was proposed for the 'toolchain-based build' thingy.

**Comment by karwa** [comment]
By manually, I mean initialising the Swift product object with a configured LLVM object.

(Sorry, GitHub doesn't have an edit button on mobile)

**Comment by ddunbar** [comment]
I agree, I think the tradeoff just depends on how generic we can make the build-script in terms of products. I'm hoping that we can get rid of a lot of the per-product duplication, and if have a single config object helps that I think it would be a win.
